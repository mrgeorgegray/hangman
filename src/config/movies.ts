/* ===============
MOVIES
all numbers and special chars removed
=============== */
const movies = [
  "a beautiful mind",
  "a charlie brown christmas",
  "a christmas story",
  "a quiet place",
  "a star is born",
  "about time",
  "ad astra",
  "alien",
  "always be my maybe",
  "american beauty",
  "anchorman",
  "annihilation",
  "apocalypto",
  "arrival",
  "arthur christmas",
  "avatar",
  "baby driver",
  "back to the future",
  "bad santa",
  "baywatch",
  "beautiful boy",
  "big hero six",
  "bird box",
  "black panther",
  "black swan",
  "black widow",
  "blade runner",
  "blockers",
  "booksmart",
  "boyhood",
  "bridesmaids",
  "bridget jones’s diary",
  "call me by your name",
  "cast away",
  "chicken run",
  "children of men",
  "chinatown",
  "christmas vacation",
  "clueless",
  "collateral",
  "crazy rich asians",
  "crazy, stupid, love",
  "dawn of the dead",
  "despicable me",
  "die hard",
  "die hard",
  "dodgeball",
  "dolittle",
  "dora and the lost city of gold",
  "dumb and dumber",
  "dumbo",
  "edge of tomorrow",
  "edward scissorhands",
  "elf",
  "eyes wide shut",
  "finding dory",
  "forgetting sarah marshall",
  "forrest gump",
  "free guy",
  "frozen",
  "game night",
  "get out",
  "ghost dog",
  "ghostbusters",
  "gifted",
  "gladiator",
  "good boys",
  "goodfellas",
  "gravity",
  "green book",
  "halloween",
  "hancock",
  "happiest season",
  "heat",
  "hereditary",
  "home again",
  "home alone",
  "how the grinch stole christmas!",
  "how to train your dragon",
  "hustlers",
  "i feel pretty",
  "inception",
  "raiders of the lost ark",
  "inside out",
  "insomnia",
  "interstellar",
  "it",
  "it happened on fifth avenue",
  "james bond: casino royale",
  "jaws",
  "john wick",
  "jojo rabbit",
  "joyeux noël",
  "kill bill",
  "kingsman: the secret service",
  "klaus",
  "knives out",
  "knocked up",
  "kong skull island",
  "kung fu panda",
  "last christmas",
  "lethal weapon",
  "liar liar",
  "lion",
  "logan",
  "love actually",
  "love actually",
  "love, simon",
  "lucky number slevin",
  "mad max",
  "mary poppins returns",
  "meet the fockers",
  "minions",
  "minority report",
  "minority report",
  "mission impossible",
  "moana",
  "monsters inc.",
  "moon",
  "moonlight",
  "mother!",
  "mulholland dr.",
  "munich",
  "my spy",
  "mystic river",
  "no country for old men",
  "noelle",
  "notting hill",
  "old school",
  "oldboy",
  "paranormal activity",
  "parasite",
  "predator",
  "pretty woman",
  "prisoners",
  "rango",
  "ready player one",
  "ring",
  "rocky",
  "rudolph, the red-nosed reindeer",
  "run lola run",
  "saving private ryan",
  "scoob!",
  "scream",
  "scrooge",
  "scrooged",
  "seven",
  "shazam!",
  "shrek",
  "shutter island",
  "silent night",
  "sin city",
  "snowpiercer",
  "sonic the hedgehog",
  "sorry to bother you",
  "soul",
  "source code",
  "southpaw",
  "spider-man",
  "split",
  "spotlight",
  "star trek the wrath of khan",
  "star wars the rise of skywalker",
  "step brothers",
  "straw dogs",
  "suicide squad",
  "superbad",
  "tag",
  "taken",
  "taxi driver",
  "tenet",
  "the accountant",
  "the big lebowski",
  "the big sick",
  "the blair witch project",
  "the bourne identity",
  "the cabin in the woods",
  "the christmas chronicles",
  "the conjuring",
  "the dark knight",
  "the departed",
  "the exorcist",
  "the favourite",
  "the fly",
  "the fugitive",
  "the game",
  "the godfather",
  "the green mile",
  "the half of it",
  "the house with a clock in its walls",
  "the human centipede",
  "the incredibles",
  "the invisible man",
  "the lego movie",
  "the lion king",
  "the longshot",
  "the martian",
  "the matrix",
  "the next three days",
  "the nice guys",
  "the night before",
  "the nightmare before christmas",
  "the notebook",
  "the other guys",
  "the others",
  "the peanut butter falcon",
  "the prestige",
  "the proposal",
  "the santa clause",
  "the secret life of pets",
  "the shawshank redemption",
  "the shining",
  "the shop around the corner",
  "the silence of the lambs",
  "the silence of the lambs",
  "the sixth sense",
  "the terminator",
  "the texas chain saw massacre",
  "the thing",
  "the transporter",
  "the usual suspects",
  "the wailing",
  "the witch",
  "there will be blood",
  "there’s something about mary",
  "thor",
  "titanic",
  "togo",
  "tomb raider",
  "total recall",
  "toy story",
  "trading places",
  "trainwreck",
  "trolls",
  "tropic thunder",
  "twelve monkeys",
  "up",
  "us",
  "v for vendetta",
  "vice",
  "wall-e",
  "wedding crashers",
  "when harry met sally",
  "whiplash",
  "wonder",
  "wreck-it ralph",
  "you’ve got mail",
  "zoolander",
];

export default movies;
