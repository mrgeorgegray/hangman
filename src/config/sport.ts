/* ===============
SPORT
all numbers and special chars removed
=============== */
const sport = [
  "aerobics",
  "archer",
  "archery",
  "arrow",
  "athlete",
  "athletics",
  "badminton",
  "ball",
  "base",
  "baseball",
  "basketball",
  "bat",
  "baton",
  "batter",
  "batting",
  "biathlon",
  "bicycle",
  "bicycling",
  "bike",
  "biking",
  "billiards",
  "bobsleigh",
  "bocce",
  "boomerang",
  "boules",
  "bow",
  "bowler",
  "bowling",
  "boxer",
  "boxing",
  "bronze medal",
  "bunt",
  "canoe",
  "canoeing",
  "catch",
  "catcher",
  "champion",
  "championship",
  "cleats",
  "club",
  "coach",
  "compete",
  "competing",
  "competition",
  "competitor",
  "crew",
  "cricket",
  "croquet",
  "cross country",
  "cue",
  "curling",
  "cycle",
  "cycling",
  "cyclist",
  "dart",
  "dartboard",
  "deadlifting",
  "decathlon",
  "defense",
  "diamond",
  "discus",
  "dive",
  "diver",
  "diving",
  "dodgeball",
  "doubleheader",
  "dugout",
  "epee",
  "equestrian",
  "equipment",
  "exercise",
  "fencing",
  "field",
  "fielder",
  "field hockey",
  "fielding",
  "figure skating",
  "fitness",
  "football",
  "forward",
  "free throw",
  "frisbee",
  "game",
  "gear",
  "geocaching",
  "goal",
  "goalie",
  "golf",
  "golfer",
  "golfing",
  "guard",
  "gym",
  "gymnast",
  "gymnastics",
  "gymnasium",
  "halftime",
  "hammer throw",
  "handball",
  "hang gliding",
  "hardball",
  "helmet",
  "heptathlon",
  "high jump",
  "hitter",
  "hockey",
  "home",
  "home plate",
  "home run",
  "home team",
  "hoop",
  "horseshoes",
  "huddle",
  "hurdle",
  "ice hockey",
  "ice rink",
  "ice skates",
  "ice skating",
  "infield",
  "infielder",
  "inline skates",
  "inning",
  "javelin",
  "jog",
  "jogger",
  "judo",
  "jump",
  "jumper",
  "jumping",
  "jump rope",
  "karate",
  "kayak",
  "kayaker",
  "kayaking",
  "kickball",
  "kneepads",
  "king fu",
  "kite",
  "lacrosse",
  "lawn bowling",
  "league",
  "long jump",
  "lose",
  "loser",
  "luge",
  "major league",
  "mallet",
  "martial arts",
  "mat",
  "medal",
  "minor league",
  "mitt",
  "mouthguard",
  "move",
  "movement",
  "net",
  "nordic skiing",
  "offense",
  "ollie",
  "orienteering",
  "out",
  "outfield",
  "outfielder",
  "paddle",
  "paddleball",
  "paddling",
  "paintball",
  "parasailing",
  "parkour",
  "pentathlon",
  "pickleball",
  "ping pong",
  "pitch",
  "pitcher",
  "play",
  "player",
  "playing",
  "playoffs",
  "pogo stick",
  "pole",
  "pole vault",
  "polo",
  "pool",
  "puck",
  "quarter",
  "quarterback",
  "quiver",
  "race",
  "racer",
  "racewalking",
  "racing",
  "racket",
  "racquetball",
  "rafting",
  "referee",
  "relay",
  "ride",
  "riding",
  "rink",
  "rock climbing",
  "roller skates",
  "roller skating",
  "row",
  "rower",
  "rowing",
  "rugby",
  "run",
  "runner",
  "running",
  "sailing",
  "score",
  "scoreboard",
  "scuba",
  "scull",
  "sculling",
  "shortstop",
  "shot put",
  "silver medal",
  "skate",
  "skating rink",
  "skeleton",
  "ski",
  "skier",
  "skiing",
  "slalom",
  "sled",
  "sledder",
  "sledding",
  "snorkling",
  "snowboard",
  "snowboarder",
  "snowboarding",
  "snowshoeing",
  "soccer",
  "softball",
  "somersault",
  "speed skating",
  "sport",
  "sportsmanship",
  "squash",
  "stadium",
  "stick",
  "strike",
  "stroke",
  "sumo wrestling",
  "surfer",
  "surfing",
  "swim",
  "swimmer",
  "swimming",
  "synchronized swimming",
  "table tennis",
  "taekwondo",
  "tag",
  "target",
  "team",
  "teammate",
  "tee",
  "telemark skiing",
  "tennis",
  "tetherball",
  "throw",
  "throwing",
  "tie",
  "toboggan",
  "track and field",
  "trampoline",
  "triathlon",
  "triathlete",
  "triple jump",
  "triple play",
  "tug of war",
  "ultramarathon",
  "umpire",
  "unicycle",
  "unicyclist",
  "uniform",
  "vaulter",
  "vaulting",
  "visiting team",
  "volley",
  "volley ball",
  "wakeboarding",
  "walk",
  "walker",
  "walking",
  "water polo",
  "waterski",
  "waterskier",
  "waterskiing",
  "weightlifter",
  "weightlifting",
  "weights",
  "wetsuit",
  "white water rafting",
  "wicket",
  "win",
  "windsurfer",
  "windsurfing",
  "winner",
  "winning",
  "wrestler",
  "wrestling",
];

export default sport;
