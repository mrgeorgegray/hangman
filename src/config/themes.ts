import { Theme } from "@emotion/react";

const baseTheme: Theme = {
  border: {
    color: "#30353B",
    radius: 3,
    width: 2,
  },
  breakpoints: {
    md: "500px",
    lg: "720px",
  },
  colors: {
    background: "#F5F5F5",
    error: "#FF0033",
    grey: "#949DA5",
    link: "#0366D6",
    linkHover: "#0387D6",
    purple: "#6E40C9",
    softError: "#FFBABA",
    softSuccess: "#9BE9A8",
    success: "#27D545",
    text: "#30353B",
    warning: "#FF7903",
  },
  fontSize: [14, 18, 22, 34, 50],
  hangman: {
    fill: "#30353B",
    stroke: "48, 53, 59",
  },
  layout: {
    maxWidth: "1000px",
  },
  space: [4, 8, 16, 32, 64, 96, 128],
  transitions: [0.15, 0.2, 0.5],
};

export const lightTheme: Theme = {
  ...baseTheme,
};

export const darkTheme: Theme = {
  ...baseTheme,
  border: {
    ...baseTheme.border,
    color: "transparent",
  },
  colors: {
    ...baseTheme.colors,
    background: "#0D1117",
    link: "#58A6FF",
    linkHover: "#ABE0FF",
    softError: "#DA3733",
    softSuccess: "#02602C",
    text: "#F5F5F5",
  },
  hangman: {
    fill: "#ADADAD",
    stroke: "245, 245, 245",
  },
};
