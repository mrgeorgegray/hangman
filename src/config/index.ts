import movies from "./movies";
import { darkTheme, lightTheme } from "./themes";

/* ===============
STARTING_CHANCES
how many guesses a user
has before the Hangman is dead
must match <Hangman /> setup.
=============== */
export const STARTING_CHANCES = 6;

/* ===============
LETTERS
determine what is displayed
on the keyboard. Make sure
you include all letters that
are in your solutions.
=============== */
// prettier-ignore
export const LETTERS = [
  "a", "b", "c", "d", "e", "f", "g", "h", "i",
  "j", "k", "l", "m", "n", "o", "p", "q", "r",
  "s", "t", "u", "v", "w", "x", "y", "z",
];

/* ===============
SOLUTIONS
@todo topics etc.
=============== */
export const SOLUTIONS = {
  movies,
};

/* ===============
THEME
@todo dark, a11y, etc.
=============== */
export const THEMES = {
  dark: darkTheme,
  light: lightTheme,
};
