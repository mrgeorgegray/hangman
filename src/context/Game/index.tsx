import React, { createContext, useContext, useReducer } from "react";
import { ThemeProvider } from "@emotion/react";
import { Immutable } from "immer";

import {
  initialState,
  curriedGameReducer,
  GameAction,
  GameState,
} from "./state";
import { THEMES } from "../../config";
import useLocalStorage from "../../hooks/useLocalStorage";

// ===============
// Setup
// ===============
export const GameDispatchContext = createContext<React.Dispatch<GameAction> | null>(
  null
);

export const GameStateContext = createContext<ReturnType<
  typeof curriedGameReducer
> | null>(null);

// ===============
// Custom hooks
// ===============
export function useGameDispatch() {
  const dispatch = useContext(GameDispatchContext);

  if (!dispatch) {
    throw new Error(
      "useGameDispatch must be used within a GameDispatchProvider."
    );
  }

  return dispatch;
}

export function useGameState() {
  const state = useContext(GameStateContext);

  if (!state) {
    throw new Error("useGameState must be used within a GameStateProvider.");
  }

  return state;
}

// ===============
// Context
// ===============
const GameContext: React.FC = ({ children }) => {
  const [gameState, setGameState] = useLocalStorage<Immutable<GameState>>(
    "gameState",
    initialState
  );
  const [state, dispatch] = useReducer(curriedGameReducer, gameState);

  React.useEffect(() => {
    setGameState(state);
  }, [state, setGameState]);

  let theme = THEMES.light;

  if (state.theme === "dark") {
    theme = THEMES.dark;
  }

  return (
    <GameDispatchContext.Provider value={dispatch}>
      <GameStateContext.Provider value={state}>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </GameStateContext.Provider>
    </GameDispatchContext.Provider>
  );
};

export default GameContext;
