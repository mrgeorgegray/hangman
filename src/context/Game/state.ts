import produce, { Draft } from "immer";

import { LETTERS, STARTING_CHANCES, SOLUTIONS, THEMES } from "../../config";
import * as utils from "./utils";

export enum GameStatus {
  NotStarted = "notStarted",
  Playing = "playing",
  Lost = "lost",
  Won = "won",
}

export type Theme = keyof typeof THEMES;

// prettier-ignore
export type GameAction =
  | { type: "giveup" }
  | { type: "guess"; payload: { letter: string }; }
  | { type: "quit" }
  | { type: "setTheme"; payload: { theme: Theme }; }
  | { type: "start" };

export interface GameState {
  chancesRemaining: number;
  guesses: string[];
  letters: string[];
  solution: string;
  solutionFormatted: string;
  status: GameStatus;
  theme: Theme;
}

export const initialState: GameState = {
  chancesRemaining: 0,
  guesses: [],
  letters: LETTERS,
  solution: "",
  solutionFormatted: "",
  status: GameStatus.NotStarted,
  theme: "light",
};

const gameReducer = (draft: Draft<GameState>, action: GameAction) => {
  switch (action.type) {
    case "start": {
      const solution = utils.sample(SOLUTIONS.movies);
      draft.chancesRemaining = STARTING_CHANCES;
      draft.guesses = [];
      draft.solution = solution;
      draft.solutionFormatted = utils.formatSolution(
        draft.solution,
        draft.guesses
      );
      draft.status = GameStatus.Playing;
      return;
    }

    case "quit": {
      draft.chancesRemaining = initialState.chancesRemaining;
      draft.guesses = initialState.guesses;
      draft.letters = initialState.letters;
      draft.solution = initialState.solution;
      draft.solutionFormatted = initialState.solutionFormatted;
      draft.status = initialState.status;
      return;
    }

    case "giveup": {
      draft.chancesRemaining = 0;
      draft.solutionFormatted = draft.solution;
      draft.status = GameStatus.Lost;
      return;
    }

    case "guess": {
      const { letter } = action.payload;

      draft.guesses.push(action.payload.letter);
      draft.solutionFormatted = utils.formatSolution(
        draft.solution,
        draft.guesses
      );
      draft.chancesRemaining = draft.solution.includes(letter)
        ? draft.chancesRemaining
        : draft.chancesRemaining - 1;

      if (draft.solution === draft.solutionFormatted) {
        draft.status = GameStatus.Won;
      }
      if (draft.chancesRemaining === 0) {
        draft.status = GameStatus.Lost;
      }
      return;
    }

    case "setTheme": {
      draft.theme = action.payload.theme;
      return;
    }

    default: {
      return;
    }
  }
};

export const curriedGameReducer = produce(gameReducer);
