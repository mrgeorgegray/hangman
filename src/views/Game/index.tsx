/** @jsxImportSource @emotion/react */
import React from "react";
import { css, useTheme } from "@emotion/react";

import { useGameDispatch, useGameState } from "../../context/Game";
import { GameStatus } from "../../context/Game/state";
import Banner from "../../components/Banner";
import Button from "../../components/Button";
import ChancesCount from "../../components/ChancesCount";
import Hangman from "../../components/Hangman";
import Letter from "../../components/Letter";
import Solution from "../../components/Solution";

const GamePage: React.FC = () => {
  const { breakpoints, space } = useTheme();
  const dispatch = useGameDispatch();
  const {
    chancesRemaining,
    guesses,
    letters,
    solution,
    solutionFormatted,
    status,
  } = useGameState();

  const handleGuess = React.useCallback(
    (letter: string) => {
      dispatch({
        type: "guess",
        payload: { letter },
      });
    },
    [dispatch]
  );

  const handleKeyup = React.useCallback(
    (event: KeyboardEvent) => {
      if (letters.includes(event.key)) {
        handleGuess(event.key);
      }
    },
    [letters, handleGuess]
  );

  const handleDispatch = (type: "new" | "quit" | "giveup") => {
    dispatch({ type });
  };

  React.useEffect(() => {
    if (status === GameStatus.Playing) {
      window.addEventListener("keyup", handleKeyup);
    } else {
      window.removeEventListener("keyup", handleKeyup);
    }

    return () => {
      window.removeEventListener("keyup", handleKeyup);
    };
  }, [status, handleKeyup]);

  return (
    <div
      css={css`
        text-align: center;

        @media (min-width: ${breakpoints.lg}) {
          text-align: left;
        }
      `}
    >
      {status === GameStatus.Lost && (
        <Banner type="error" message="Game over :-( ">
          <Button
            type="plain"
            onClick={() => handleDispatch("new")}
            text="New Game"
          />
        </Banner>
      )}
      {status === GameStatus.Won && (
        <Banner type="success" message="Congratulations! You Won :-)">
          <Button
            type="plain"
            onClick={() => handleDispatch("new")}
            text="New Game"
          />
        </Banner>
      )}

      <div
        css={css`
          margin-top: ${space[2]}px;

          @media (min-width: ${breakpoints.lg}) {
            float: left;
            width: 48%;
          }
        `}
      >
        <Hangman chancesRemaining={chancesRemaining} />
      </div>
      <div
        css={css`
          margin-top: ${space[2]}px;

          @media (min-width: ${breakpoints.lg}) {
            float: right;
            margin-top: ${space[4]}px;
            width: 48%;
          }
        `}
      >
        <Solution text={solutionFormatted} status={status} />
        <ChancesCount chancesRemaining={chancesRemaining} />
        <div
          css={css`
            margin-bottom: ${space[3]}px;

            @media (min-width: ${breakpoints.lg}) {
              margin-left: -${space[2] + 2}px;
              max-width: 440px;
            }
          `}
        >
          {letters.flatMap((letter) => {
            const isUsed = guesses.includes(letter);
            const isDisabled = status !== GameStatus.Playing || isUsed;
            const isCorrect = solution?.includes(letter);
            const shouldHighlight = isUsed && isDisabled;

            return (
              <Letter
                aria-controls="chancesRemaining"
                key={letter}
                onClick={handleGuess}
                letter={letter}
                isDisabled={isDisabled}
                isCorrect={isCorrect}
                shouldHighlight={shouldHighlight}
              />
            );
          })}
        </div>
        <nav>
          {status === GameStatus.Playing && (
            <div>
              <Button
                type="warning"
                onClick={() => handleDispatch("giveup")}
                text="Give up!"
              />
              <Button
                type="error"
                onClick={() => handleDispatch("quit")}
                text="Quit Game"
              />
            </div>
          )}
        </nav>
      </div>
    </div>
  );
};

export default GamePage;
