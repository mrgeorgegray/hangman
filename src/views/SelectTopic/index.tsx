/** @jsxImportSource @emotion/react */
import { css, useTheme } from "@emotion/react";

import { useGameDispatch } from "../../context/Game";
import { Topic, topicList } from "../../context/Game/state";
import Button from "../../components/Button";
import { capitaliseFirstLetter } from "../../utils";

const SelectTopic = () => {
  const { space } = useTheme();
  const dispatch = useGameDispatch();

  const handleSetTopic = (topic: Topic) => {
    dispatch({ type: "setTopicAndStart", payload: { topic } });
  };

  return (
    <div
      css={css`
        text-align: center;
      `}
    >
      <p
        css={css`
          margin-bottom: ${space[3]}px;
          margin-top: ${space[4]}px;
        `}
      >
        Choose a topic
      </p>
      {topicList.map((topic) => (
        <div
          key={topic}
          css={css`
            margin-bottom: ${space[2]}px;
          `}
        >
          <Button
            type="plain"
            size="large"
            onClick={() => handleSetTopic(topic)}
            text={capitaliseFirstLetter(topic)}
          />
        </div>
      ))}
    </div>
  );
};

export default SelectTopic;
