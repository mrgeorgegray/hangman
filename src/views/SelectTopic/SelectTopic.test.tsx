import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { ThemeProvider } from "@emotion/react";

import SelectTopic from ".";
import { GameDispatchContext, GameStateContext } from "../../context/Game";
import { GameAction, initialState } from "../../context/Game/state";
import { THEMES } from "../../config";

describe("<SelectTopic />", () => {
  const mockDispatch: React.Dispatch<GameAction> = jest.fn();

  const buildSubject = () =>
    render(
      <GameDispatchContext.Provider value={mockDispatch}>
        <GameStateContext.Provider value={initialState}>
          <ThemeProvider theme={THEMES.light}>
            <SelectTopic />
          </ThemeProvider>
        </GameStateContext.Provider>
      </GameDispatchContext.Provider>
    );

  it("renders", () => {
    const { container } = buildSubject();

    expect(container).toMatchSnapshot();
  });

  it("handles selecting a topic", () => {
    const { getByText } = buildSubject();

    fireEvent.click(getByText("Words"));

    expect(mockDispatch).toHaveBeenCalledWith({
      type: "setTopicAndStart",
      payload: { topic: "words" },
    });
  });
});
