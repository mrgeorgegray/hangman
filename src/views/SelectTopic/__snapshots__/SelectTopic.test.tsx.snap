// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectTopic /> renders 1`] = `
.emotion-0 {
  text-align: center;
}

.emotion-1 {
  margin-bottom: 32px;
  margin-top: 64px;
}

.emotion-2 {
  margin-bottom: 16px;
}

.emotion-3 {
  background-color: #F5F5F5;
  border: 2px solid #30353B;
  border-radius: 3px;
  color: #30353B;
  font-size: 22px;
  outline: none;
  padding: 16px 24px;
  margin: 0 4px;
  -webkit-transition: all 0.15s ease-in;
  transition: all 0.15s ease-in;
}

.emotion-3:hover {
  color: #F5F5F5;
  cursor: pointer;
  background-color: #30353B;
  border-color: #30353B;
}

<div>
  <div
    class="emotion-0"
  >
    <p
      class="emotion-1"
    >
      Choose a topic
    </p>
    <div
      class="emotion-2"
    >
      <button
        class="emotion-3"
      >
        Movies
      </button>
    </div>
    <div
      class="emotion-2"
    >
      <button
        class="emotion-3"
      >
        Sport
      </button>
    </div>
    <div
      class="emotion-2"
    >
      <button
        class="emotion-3"
      >
        Words
      </button>
    </div>
  </div>
</div>
`;
